File: macos-reverse-engineering-10weeks/code/week02/macho_parser.cpp
// 02_macho_parser.cpp — Đọc header Mach-O + nhận diện fat/universal binary
// Mục tiêu: magic bytes, cputype (arm64/x86_64), filetype, số load commands.
// An toàn: chỉ ĐỌC header, không thực thi.  g++ -std=c++17 -O2 macho_parser.cpp -o demo
#include <iostream>
#include <fstream>
#include <vector>
#include <cstdint>

uint32_t rd32(const std::vector<uint8_t>& d, size_t o){ return (uint32_t)d[o]|(d[o+1]<<8)|(d[o+2]<<16)|(d[o+3]<<24); }

int main(int argc, char** argv) {
    if (argc < 2) { std::cout << "Dùng: ./demo <file>\n"; return 1; }
    std::ifstream f(argv[1], std::ios::binary);
    std::vector<uint8_t> d((std::istreambuf_iterator<char>(f)), {});
    if (d.size() < 4) return 1;

    uint32_t magic = rd32(d, 0);
    if (magic == 0xCAFEBABE) {
        std::cout << "FAT / UNIVERSAL binary (" << rd32(d, 4) << " kiến trúc)\n";
    } else if (magic == 0xFEEDFACF) {
        std::cout << "Mach-O 64-bit (little-endian)\n";
    } else if (magic == 0xFEEDFACE) {
        std::cout << "Mach-O 32-bit (little-endian)\n";
    } else if (magic == 0xCFFAEDFE) {
        std::cout << "Mach-O 64-bit (big-endian)\n";
    } else {
        std::cout << "Không phải Mach-O (magic 0x" << std::hex << magic << std::dec << ")\n";
        return 0;
    }

    if (d.size() >= 32) {
        uint32_t cpu = rd32(d, 4);
        uint32_t filetype = rd32(d, 12);
        uint32_t ncmds = rd32(d, 16);
        std::cout << "  cputype: " << (cpu == 0x0100000C ? "ARM64" : cpu == 0x01000007 ? "x86_64" : "khác") << "\n";
        std::cout << "  filetype: " << filetype << "\n";
        std::cout << "  load commands: " << ncmds << "\n";
    }
    std::cout << "Kích thước: " << d.size() << " bytes\n";
    return 0;
}

File: macos-re-10weeks/code/week02/macho_parser.py
#!/usr/bin/env python3
# ============================================================
# macho_parser.py — Parse header Mach-O (dữ liệu giả/minh hoạ)
# Mục tiêu: đọc và in header Mach-O từ một file nhị phân giả.
# Đầu vào: file (mặc định tự tạo dữ liệu giả nếu không có).
# Đầu ra: header, load commands, segments.
# An toàn: chỉ phân tích file do chính bạn tạo; không đọc mẫu malware.
# ============================================================
import struct
import sys

# Magic numbers Mach-O (minh hoạ)
MH_MAGIC_64 = 0xFEEDFACF
MH_MAGIC_32 = 0xFEEDFACE
FAT_MAGIC = 0xCAFEBABE

# CPU types (một phần)
CPU_ARCH_ABI64 = 0x01000000
CPU_TYPE_X86_64 = 0x01000007
CPU_TYPE_ARM64 = 0x0100000C

# File types
MH_EXECUTE = 0x2
MH_DYLIB = 0x6
MH_BUNDLE = 0x8

LOAD_CMDS = {
    0x1: "LC_SEGMENT",
    0x19: "LC_SEGMENT_64",
    0xC: "LC_LOAD_DYLIB",
    0xD: "LC_ID_DYLIB",
    0x1D: "LC_CODE_SIGNATURE",
    0x80000028: "LC_MAIN",
}


def make_fake_header():
    """Tạo 32 byte header Mach-O 64-bit giả cho demo."""
    # magic, cputype, cpusubtype, filetype, ncmds, sizeofcmds, flags, reserved
    return struct.pack(
        "<IiiIIIII",
        MH_MAGIC_64,
        CPU_TYPE_ARM64,
        0,
        MH_EXECUTE,
        3,
        96,
        0,
        0,
    )


def parse(data: bytes):
    if len(data) < 32:
        print("Quá ngắn để là Mach-O 64-bit.")
        return
    magic, cputype, cpusub, filetype, ncmds, sizeofcmds, flags, _ = struct.unpack(
        "<IiiIIIII", data[:32]
    )
    print(f"magic        : 0x{magic:08X} ({'64-bit' if magic == MH_MAGIC_64 else '32-bit' if magic == MH_MAGIC_32 else 'không phải Mach-O'})")
    print(f"cputype      : {cputype} ({'arm64' if cputype == CPU_TYPE_ARM64 else 'x86_64' if cputype == CPU_TYPE_X86_64 else '?'})")
    ftype = "MH_EXECUTE" if filetype == MH_EXECUTE else "other"
    print(f"filetype     : {filetype} ({ftype})")
    print(f"ncmds        : {ncmds}")
    print(f"sizeofcmds   : {sizeofcmds}")
    print("load commands (minh hoạ):")
    for i in range(min(ncmds, 4)):
        print(f"  LC_{i}: {LOAD_CMDS.get(0x19, 'LC_SEGMENT_64')}")


def main():
    if len(sys.argv) > 1:
        data = open(sys.argv[1], "rb").read()
    else:
        data = make_fake_header()
        print("(dùng header giả tự tạo)\n")
    parse(data)


if __name__ == "__main__":
    main()
